types/big-json/index.d.ts

```diff
@@ -0,0 +1,29 @@
+// Type definitions for big-json 3.2
+// Project: https://github.com/DonutEspresso/big-json
+// Definitions by: PCOffline <https://github.com/PCOffline>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+
+/// <reference types="node" />
+
+import type { Stream } from 'node:stream';
+
+export function createParseStream(): Stream;
+export function createStringifyStream(opts: { body: object }): Stream;
+
+export function parse(opts: { body: string | Buffer }): Promise<object | ReadonlyArray<object>>;
+export function parse(
+    opts: { body: string | Buffer },
+    callback: (result: object | ReadonlyArray<object>) => void,
+): void;
+
+export function stringify(opts: { body: object }): Promise<string>;
+export function stringify(opts: { body: object }, callback: (result: string) => void): void;
+
+declare const exported: {
+    createParseStream: typeof createParseStream;
+    createStringifyStream: typeof createStringifyStream;
+    parse: typeof parse;
+    stringify: typeof stringify;
+};
+
+export default exported;
```

Comment:
why do you think this is ES6-kind module with export default?
https://[unpkg.com/big-json@3.2.0/lib/index.js](https://unpkg.com/big-json@3.2.0/lib/index.js)
this seems to be old, plain CJS module

Reply:
Fixed!

Reply:
> According to the [docs](https://www.typescriptlang.org/docs/handbook/declaration-files/templates/module-d-ts.html), the ES6 syntax, in this case, is equivalent to the CommonJS behaviour, as far as I can tell.

Please review the rest of the linked page:

> Note that using export default in your .d.ts files requires [esModuleInterop: true](https://www.typescriptlang.org/tsconfig#esModuleInterop) to work. If you can’t have esModuleInterop: true in your project, ***such as when you’re submitting a PR to Definitely Typed, you’ll have to use the export= syntax instead.***

the CJS module uses only named functions, so it could be expressed as explicit exports, so:
```ts
/// <reference types="node" />

import type { Stream } from 'node:stream';

export function createParseStream(): Stream;
export function createStringifyStream(opts: { body: object }): Stream;

export function parse(opts: { body: string | Buffer }): Promise<object | ReadonlyArray<object>>;
export function parse(
    opts: { body: string | Buffer },
    callback: (result: object | ReadonlyArray<object>) => void,
): void;

export function stringify(opts: { body: object }): Promise<string>;
export function stringify(opts: { body: object }, callback: (result: string) => void): void;
```
usage in tests:
```ts
import * as bigJson from 'big-json';
```

and yes, users of this DT module, when published, could (when using `esModuleInterop`) use ESM syntax by default.

Reply:
Alright, thanks! I'll fix it in a few hours

Reply:
On second thought, if I merely use named-export, the first test will fail. I'll emphasize that both code snippets were run on **vanilla JS** and worked.

Reply:
Vanilla JS meaning node? Node is going to give the `module.exports` object back and that's what you're seeing here.

`export default` is effectively never the answer when describing `cjs` files.

I'm having trouble understanding what you're saying the problem is here. As far as I'm aware, what @peterblazejewicz wrote is correct and these can all just be explicit named exports with no default, and that's correct.

Reply:
I'm under impression we're now discussing how to load and use CJS with ESM module support in modern Node runtime (plain JS feature in Node/web), not how to write a ESM compatible (via ESModuleInterop) TS declaration file. I've linked TS handbook as a reference. Users of DT module will be able to use `import bigJson from 'big-json';` syntax be assured.

Reply:
I don't exactly understand you. I mean that by using merely named exports, the DT module misrepresents the export behaviour of `big-json`. I am uncertain whether this would affect the TS compiler when trying to use default imports with `big-json` (without `esModuleInterop`) – I am prone to believe it won't – but that's not my issue, it's merely the misrepresentation that concerns me.

Reply:
What I'm saying is that the named exports are not a misrepresentation; I know the rules are a little confusing but when you are just describing a bunch of properties on `module.exports` and `module.exports` is not a function/class/whatever, you represent that in a `d.ts` file with named exports.

If you test it without the extra `default` stuff, you get it working both ways:

![image](https://github.com/DefinitelyTyped/DefinitelyTyped/assets/5341706/f587857e-18e4-4c61-a67a-1d227095e9f1)

Reply:
I see. Did you use the `export = ` syntax or just used named exports? Because both don't seem to achieve what you're showing in the pic

Reply:
Just the named exports. I copied and pasted this PR and removed `exported` and `export default`.

Reply:
Weird, for me the tests are failing:
![image](https://github.com/DefinitelyTyped/DefinitelyTyped/assets/21259184/001d08c8-d2df-4148-bf28-7bf41221e6d1)
![image](https://github.com/DefinitelyTyped/DefinitelyTyped/assets/21259184/ff90d43a-ae91-4099-b5db-e453f39331f5)

Reply:
@andrewbranch Do you know what's going on here? Am I fundamentally forgetting something?

Reply:
I noticed your `moduleResolution` & `module` are different than the `tsconfig.json` in the PR, although when I change them to match yours, it doesn't recognise `big-json` as a module at all.

Reply:
No, you’re right. @PCOffline just needed to update the test file to one of these

```diff
- import bigJson from "big-json"
+ import bigJson = require("big-json");
// or
+ import * as bigJson from "big-json";
```

The default import will only work with `esModuleInterop` enabled.

Reply:
Ugh, I didn't even notice that it hadn't written `require`. That's the actual thing to do here.

Reply:
alright it works now!